docs/DynamicCasting.md

```diff
@@ -0,0 +1,513 @@
+
+* Class downcasts: If `C` is a subclass of `SuperC` and `sc` is an instance of `SuperC`, then `sc is C` will be true iff `sc` is actually an instance of `C`.  When this "downcast" does succeed, it does not change the representation.
+
+* Objective-C class casting:  The rules above also apply when one of the classes in question is defined with the `@objc` attribute or inherits from an Objective-C class.
+
+* Class casts to AnyObject: Any class reference can be cast to `AnyObject` and then cast back to the original type.  See "AnyObject" below.
+
+* If a struct or enum type conforms to `_ObjectiveCBridgeable`, then classes of the associated bridging type can be cast to the struct or enum type and vice versa.  See "The `_ObjectiveCBridgeable` Protocol" below.
+
+Invariants:
+* For any two class types `C1` and `C2`: `c is C1 && c is C2` implies that `(c as! C1) is C2`
+* For any class type `C`: `c is C` implies that `(c as! AnyObject) as! C === c`
+
+## Structs and Enums
+
+You cannot cast between distinct struct or enum types.
+More formally:
+
+* If `S` is any struct or enum type, `s is S == true`, and `T` is any type, then `s is T` iff `S.self == T.self`.
+
+## Tuples
+
+Casting from a tuple type T1 to a tuple type T2 will succeed iff the following hold:
+* T1 and T2 have the same number of elements
+* If an element in T1 has a label, the corresponding element in T2 has the same label
+* Each element of T1 can be individually cast to the corresponding type of T2
+
+## Functions
+
+Casting from a function type F1 to a function type F2 will succeed iff the following hold:
+* The two types have the same number of arguments
+* Corresponding arguments have identical types
+* The return types are identical
+* If F1 is a throwing function type, then F2 must be a throwing function type.  If F1 is not throwing, then F2 may be a throwing or non-throwing function type.
+
+Note that it is _not_ sufficient for argument and return types to be castable; they must actually be identical.
+
+## Optionals
+
+Casting to and from optional types will transparently unwrap optionals as much as necessary, including nested optional types.
+
+* If `T` and `U` are any two types, and `t` is an instance of `T`, then `.some(t) is Optional<U> == t is U`
+* Optional injection:  if `T` and `U` are any two types, `t` is a non-nil instance of `T`, then `t is Optional<U> == t is U`
+* Optional projection:  if `T` and `U` are any two types, and `t` is an instance of `T`, then `.some(t) is U == t is U`
+* Nil Casting: if `T` and `U` are any two types, then `Optional<T>.none is Optional<U> == true`
+
+Note: For "Optional Injection" above, the requirement that `t` is a non-nil instance of `T` implies that either `T` is not an `Optional` or that `T` is `Optional` and `t` has the form `.some(u)`
+
+XXX TODO: Is it correct to say that in general `as!` commutes with optional injection?
```

Comment:
I don't think so. I've seen the following terminology used:

1. `A` commutes with `B` iff `AB = BA`.
2. `★` is a commutative operation iff `a ★ b = b ★ a` for all `a`, `b`.

This doesn't fall into either of those styles.

Reply:
Thanks.  I've removed this.